sdk/core/Azure.Core/src/ResourceIdentifier.cs

```diff
@@ -0,0 +1,429 @@
+            IsChild = isChild;
+            if (parent is null)
+                _stringValue = RootStringValue;
+        }
+
+        private ResourceIdentifier(ResourceIdentifier parent, string resourceTypeName, string resourceName)
+            : this(parent, ChooseResourceType(resourceTypeName, parent), resourceName, true)
+        {
+        }
+
+        private ResourceIdentifier(ResourceIdentifier parent, string providerNamespace, string resourceTypeName, string resourceName)
+            : this(parent, new ResourceType(providerNamespace, resourceTypeName), resourceName, false)
+        {
+        }
+
+        private void Init(ResourceIdentifier? parent, ResourceType resourceType, string resourceName)
+        {
+            if (parent is not null)
+            {
+                Provider = parent.Provider;
+                SubscriptionId = parent.SubscriptionId;
+                Location = parent.Location;
+                ResourceGroupName = parent.ResourceGroupName;
+            }
+
+            if (resourceType == SubscriptionResourceType)
+            {
+                if (!Guid.TryParse(resourceName, out _))
+                    throw new ArgumentOutOfRangeException(nameof(resourceName), $"The GUID for subscription is invalid {resourceName}.");
+                SubscriptionId = resourceName;
+            }
+
+            if (resourceType.GetLastType() == LocationsKey)
+                Location = resourceName;
+
+            if (resourceType == ResourceGroupResourceType)
+                ResourceGroupName = resourceName;
+
+            if (resourceType == ProviderResourceType)
+                Provider = resourceName;
+        }
+
+        private static ResourceType ChooseResourceType(string resourceTypeName, ResourceIdentifier parent)
+        {
+            if (resourceTypeName.Equals(ResourceGroupsLowerKey, StringComparison.OrdinalIgnoreCase))
+                return ResourceGroupResourceType;
+
+            //subscriptions' type is Microsoft.Resources/subscriptions only when its parent is Tenant
+            if (resourceTypeName.Equals(SubscriptionsKey, StringComparison.OrdinalIgnoreCase) && parent.ResourceType == TenantResourceType)
+                return SubscriptionResourceType;
+
+            return parent.ResourceType.AppendChild(resourceTypeName);
+        }
+
+        private static ResourceIdentifier Create(string resourceId)
+        {
+            Argument.AssertNotNullOrEmpty(resourceId, nameof(resourceId));
+
+            if (resourceId[0] != Separator)
+                throw new ArgumentOutOfRangeException(nameof(resourceId), "The ResourceIdentifier must start with '/'.");
+
+            if (resourceId.Length == 1)
+                return Root;
+
+            string[] parts = resourceId.Split(Separator);
+            if (parts.Length < 3)
+                throw new ArgumentOutOfRangeException(nameof(resourceId), "The ResourceIdentifier is too short it must have at least 2 path segments.");
```

Comment:
Is the ResourceIdentifier format documented somewhere?  It might be helpful to users if we provide a link to this in the comments, or include a description of what input format the type expects in the docstrings.